### PR TITLE
feat: add OverlapPatchEmbed module of SVTR

### PR DESCRIPTION
**Pull request by willpat1213**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmocr/blob/main/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmocr/blob/main/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmocr/models/textrecog/backbones/svtr.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# Modified from <https://arxiv.org/abs/2205.00159>
+# Adapted from <https://github.com/PaddlePaddle/PaddleOCR>
+from typing import Callable
+
+import torch.nn as nn
+from mmengine.model import BaseModule
+
+
+class Conv_BN_Layer(BaseModule):
```
Comment (Mountchicken) [inline mmocr/models/textrecog/backbones/svtr.py:10]: We can use [ConvMoudle](https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/conv_module.py#L18) in mmcv instead

File: mmocr/models/textrecog/backbones/svtr.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        embed_dims (int): The dimensions of embedding.
+        num_layers (int, optional): Number of Conv_BN_Layer. Defaults to 2.
+    """
+
+    def __init__(self,
+                 img_size=[32, 100],
+                 in_channels=3,
+                 embed_dims=768,
+                 num_layers=2,
+                 init_cfg=None):
```
Comment (Mountchicken) [inline mmocr/models/textrecog/backbones/svtr.py:76]: Missing typings and default values. We can take [this](https://github.com/open-mmlab/mmocr/blob/dev-1.x/mmocr/models/textrecog/encoders/sar_encoder.py#L20-L47) as an example

File: mmocr/models/textrecog/backbones/svtr.py (hunk 1 above)
Comment (willpat1213) [reply mmocr/models/textrecog/backbones/svtr.py]: Thanks for your reminding, I have used ConvModule instead.

File: mmocr/models/textrecog/backbones/svtr.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+        super().__init__(init_cfg=init_cfg)
+        """
+        num_patches = (img_size[1] // (2 ** num_layers)) * \
+                      (img_size[0] // (2 ** num_layers))
+        """
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:33]: Delete. BTW, triple quotes in python are used to denote docstrings instead of comments

**Review by gaotongxiao — COMMENTED** [review]
Need some polishment after all the SVTR modules are implemented

File: mmocr/models/textrecog/backbones/svtr.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        self.img_size = img_size
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:34]: Is `img_size` still necessary?

File: mmocr/models/textrecog/backbones/svtr.py (hunk 3 above)
Comment (willpat1213) [reply mmocr/models/textrecog/backbones/svtr.py:33]: Because I need the object `num_patches` to construct the model in the end, I used the triple quotes to prevent forgetting.

File: mmocr/models/textrecog/backbones/svtr.py (hunk 4 above)
Comment (willpat1213) [reply mmocr/models/textrecog/backbones/svtr.py:34]: Yeah, `img_size` is used to assert input image size matches the model or not in the `forward`.

File: mmocr/models/textrecog/backbones/svtr.py (hunk 3 above)
Comment (gaotongxiao) [reply mmocr/models/textrecog/backbones/svtr.py:33]: I'd suggest removing it if `img_size` is just used to verify the dimension irrelevant to a successful forward of this module. Modules are better implemented with no assumption on their runtime context and should not assert something outside their scope.

File: mmocr/models/textrecog/backbones/svtr.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        embed_dims (int): The dimensions of embedding. Defaults to 768.
+        num_layers (int, optional): Number of Conv_BN_Layer. Defaults to 2.
+    """
+
+    def __init__(self,
+                 img_size=[32, 100],
+                 in_channels=3,
+                 embed_dims=768,
+                 num_layers=2,
+                 init_cfg=None):
```
[the remark this answers is not in the document]
Comment (willpat1213) [reply mmocr/models/textrecog/backbones/svtr.py]: In the latest version, I add the necessary type hint.

File: mmocr/models/textrecog/backbones/svtr.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+import torch.nn as nn
+from mmcv.cnn import ConvModule
+from mmengine.model import BaseModule
+
+
+class OverlapPatchEmbed(BaseModule):
+    """Image to the progressive overlapping Patch Embedding.
+
+    Args:
+        input_size (int | tuple): The size of input, which will be used to
```
[the remark this answers is not in the document]
Comment (willpat1213) [reply mmocr/models/textrecog/backbones/svtr.py]: Updated in the latest version.

File: mmocr/models/textrecog/backbones/svtr.py [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+        self.proj_drop = nn.Dropout(proj_drop)
+        self.HW = HW
+        if HW is not None:
+            H, W = HW[0], HW[1]
+            self.N = H * W
+            self.C = embed_dims
+        if mixer == 'Local' and HW is not None:
+            hk = local_k[0]
+            wk = local_k[1]
+            mask = torch.ones([H * W, H + hk - 1, W + wk - 1], dtype='float32')
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:169]:
```suggestion
            mask = torch.ones([H * W, H + hk - 1, W + wk - 1], dtype=torch.float)
```
`'float32'` doesn't work on PyTorch. Your unit tests should cover as many lines as possible to find out these error at the very beginning. We have a script to help you test if your new code snippets fit our unit test and docstring coverage requirements.

Here is a brief description of this script. Make sure you run this script and pass all the tests before you finalize your commits.

## Description

This script automatically tests the unit test and docstring coverage of all added/modified python files in `mmocr/` in the latest commit against a specific branch (defaults to `origin/refactor_dev`).
Under certain circumstances, you might not want some files to be tested. You can add its relative path to `.dev_scripts/covignore.cfg`. Regular expression is also supported.

## Usage
To locally test your changes against `origin/refactor_dev`:

```bash
.dev_scripts/diff_coverage_test.sh
```

Test against `open-mmlab/main`:

```bash
.dev_scripts/diff_coverage_test.sh open-mmlab main
```

Comment (willpat1213) [reply mmocr/models/textrecog/backbones/svtr.py:169]: Had fixed in the next commit.

File: mmocr/models/textrecog/backbones/svtr.py [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+        attn = F.softmax(attn, dim=-1)
+        attn = self.attn_drop(attn)
+
+        x = attn.matmul(v).permute(0, 2, 1, 3).reshape(-1, N, C)
+        x = self.proj(x)
+        x = self.proj_drop(x)
+        return x
+
+
+class Mlp(BaseModule):
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:209]:
```suggestion
class MLP(BaseModule):
```

File: mmocr/models/textrecog/backbones/svtr.py [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+    """The MLP block.
+
+    Args:
+        in_features (int): The input features.
+        hidden_features (int, optional): The hidden features.
+            Defaults to None.
+        out_features (int, optional): The output features.
+            Defaults to None.
+        drop (float, optional): cfg of dropout function. Defaults to 0.0.
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:218]: P2: docstr for init_cfg

File: mmocr/models/textrecog/backbones/svtr.py [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+        local_mixer (Tuple[int ,int], optional): Local window size.
+            Defaults to [7, 11].
+        HW (Tuple[int, int], optional): The size of [H, W].
+            Defaults to [8, 25].
+        mlp_ratio (float, optional): The ratio of hidden features to input.
+            Defaults to 4.0.
+        qkv_bias (bool, optional): Whether a additive bias is required.
+            Defaults to False.
+        qk_scale (float, optional): A scaling factor. Defaults to None.
+        drop (float, optional): _description_. Defaults to 0..
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:268]: P2: To be done

File: mmocr/models/textrecog/backbones/svtr.py [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+        attn_drop (float, optional): A Dropout layer. Defaults to 0.0.
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:269]: P2: A Dropout layer for what

File: mmocr/models/textrecog/backbones/svtr.py [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+        embed_dims (int): Number of character components.
+        num_heads (int, optional): Number of heads. Defaults to 8.
+        mixer (str, optional): The mixer type. Defaults to 'Global'.
+        HW (Tuple[int, int], optional): Number of H x W. Defaults to [8, 25].
+        local_k (Tuple[int, int], optional): Window size. Defaults to [7, 11].
+        qkv_bias (bool, optional): Whether a additive bias is required.
+            Defaults to False.
+        qk_scale (float, optional): A scaling factor. Defaults to None.
+        attn_drop (float, optional): A Dropout layer. Defaults to 0.0.
+        proj_drop (float, optional): A Dropout layer. Defaults to 0.0.
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:137]: P2: A dropout layer for what

File: mmocr/models/textrecog/backbones/svtr.py [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+        mlp_ratio (float, optional): The ratio of hidden features to input.
+            Defaults to 4.0.
+        qkv_bias (bool, optional): Whether a additive bias is required.
+            Defaults to False.
+        qk_scale (float, optional): A scaling factor. Defaults to None.
+        drop (float, optional): _description_. Defaults to 0..
+        attn_drop (float, optional): A Dropout layer. Defaults to 0.0.
+        drop_path (_type_, optional): The probability of drop path.
+            Defaults to 0.0.
+        pernorm (bool, optional): Mixing layer before norm. Defaults to True.
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:272]:
```suggestion
        prenorm (bool, optional): Whether to place the mixing block before norm. Defaults to True.
```

File: mmocr/models/textrecog/backbones/svtr.py [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+        return x
+
+
+class MixingBlock(BaseModule):
+    """The Mixing block.
+
+    Args:
+        embed_dims (int): Number of character components.
+        num_heads (int): Number of heads
+        mixer (str, optional): The mixer type. Defaults to 'Global'.
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:258]: P2: Options are...

File: mmocr/models/textrecog/backbones/svtr.py [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+        if self.prenorm:
+            x = self.norm1(x + self.drop_path(self.mixer(x)))
+            x = self.norm2(x + self.drop_path(self.mlp(x)))
+        else:
+            x = x + self.drop_path(self.mixer(self.norm1(x)))
+            x = x + self.drop_path(self.mlp(self.norm2(x)))
+        return x
+
+
+class DownSample(BaseModule):
```
Comment (gaotongxiao) [inline mmocr/models/textrecog/backbones/svtr.py:333]: Where will it be used?

**Review by gaotongxiao — COMMENTED** [review]
Comments denoted as P2 means not rushy to fix.
